### PR TITLE
Fix TextBox not invalidating properly

### PR DESCRIPTION
**Pull request by default0**
If a user binds the Current.ValueChanged-Event of a TextBox and changes
the text of the textbox inside the ValueChanged-Handler either by
directly assigning to TextBox.Current.Value or TextBox.Text the TextBox
will not show the newly assigned text until another change is made to
the contents or caret of the TextBox by either User Input or Code run
on a different Update.

The reason this happens is because regardless of if Current.Value or the text of the TextBox has changed, the layout will get validated after updateCursorAndLayout() has finished executing - however, updateCursorAndLayout() also happens to be the trigger for the ValueChanged-Event. Thus, if changes to the TextBox Text or Current.Value are made at this point, the (after calling the callback now outdated) Layout will still be assumed valid.

I fixed this by validating the layout first and then calling updateCursorAndLayout() after (just switching the order) - now if changes to TextBox.Text or Current.Value are made in a ValueChanged-Handler, the calls to invalidate the Layout made by these actions successfully invalidate the Layout instead of being overridden by the later call to cursorAndLayout.Validate() inside TextBox.UpdateAfterChildren().

This also adds a test to verify this behaviour.

### CODE REVIEW

File: osu.Framework.Tests/Visual/UserInterface/TestSceneTextBox.cs [hunk 1]
```diff
@@ -182,6 +195,30 @@ public void TestNumbersOnly()
...
+                });
+                overrideInputBox.Current.BindValueChanged(vce =>
+                {
+                    if (vce.NewValue != @"Input overridden!")
+                        overrideInputBox.Text = @"Input overridden!";
+                });
+            });
+
+            AddStep(@"set some text", () => overrideInputBox.Text = "smth");
+            AddAssert(@"verify display state", () => overrideInputBox.TextFlow.Children.Count == "Input overridden!".Length);
```
[the remark this answers is not in the document]
Comment (default0) [reply osu.Framework.Tests/Visual/UserInterface/TestSceneTextBox.cs]: Done :-) Thanks for the hint!

**Comment by default0** [comment]
> 
> 
> Honestly sounds like the whole `textUpdateSchedule` thing should just be removed and use `ScheduleAfterChildren` instead, it doesn't seem to be doing any special behaviour compared to regular schedules.

Generally not super sure on this, sounds like it'd end up being a minor refactor of the TextBox class as a whole - I just wanted to get out a fix+test for a particular problem - should maybe be a separate issue to investigate if removing it makes sense/refactoring the TextBox class a bit?

**Comment by bdach** [comment]
This feels like a very flimsy fix, though. To me validating layout before actually updating it reads like a bug.

**Comment by frenzibyte** [comment]
> Generally not super sure on this, sounds like it'd end up being a minor refactor of the TextBox class as a whole - I just wanted to get out a fix+test for a particular problem - should maybe be a separate issue to investigate if removing it makes sense/refactoring the TextBox class a bit?

should just be a `textUpdateSchedule.Add(` -> `ScheduleAfterChildren(`

**Comment by default0** [comment]
> 
> should just be a `textUpdateSchedule.Add(` -> `ScheduleAfterChildren(`

Just tried this locally, it doesn't cause any test failures and for the most part textboxes still behave the same, however the caret positioning in the case of the overriding textbox gets broken by this change (the caret displays behind the 1st character after the text override, instead of at its original position, it fixes itself if you move it though).

> 
> 
> This feels like a very flimsy fix, though. To me validating layout before actually updating it reads like a bug.

Order of operations tends to be extremely important (and easy to get wrong) when dealing with events & callbacks. I generally agree on how it "feels", but matter of fact is that the old order of operations (validating layout after the update) caused a bug.
There's a lot of other decisions one could be making to potentially address this issue:

- Not delaying ValueChanged until the new text gets rendered
- Always updating the display state immediately when the text changes (and doing offset calculations for the caret "by hand")
- Creating a full separation between what the textbox displays and what its internal value is and auto-invalidating on mismatch in Update
- Probably a bunch of other approaches

The issue with most (all?) of them is that they are going to be breaking compatibility in various hard-to-detect-for-users ways and I did not intend to refactor the TextBox class just to fix a single issue in a specific scenario that I encountered with it. I think refactoring the class to make it work better in general should be separate from fixing specific bugs.

**Comment by bdach** [comment]
> Order of operations tends to be extremely important (and easy to get wrong) when dealing with events & callbacks. I generally agree on how it "feels", but matter of fact is that the old order of operations (validating layout after the update) caused a bug.

I am well aware of this.

> The issue with most (all?) of them is that they are going to be breaking compatibility in various hard-to-detect-for-users ways and I did not intend to refactor the TextBox class just to fix a single issue in a specific scenario that I encountered with it. I think refactoring the class to make it work better in general should be separate from fixing specific bugs.

The thing is though that in my view accepting this change amounts to accepting technical debt that will have to be repaid at some point down the line. If the control flow in the text box can be so confusing that this is an issue, then we might run into something else or other like this again. I don't see the approach of "don't ever fix bugs by refactoring" as valid. Tacking on cryptic fixes (like this one invariably is) onto hard-to-follow classes is not a recipe for success in my book.

**Comment by default0** [comment]
I guess I'll just wait for someone else to feel like refactoring this class at some point and build workarounds in my user-code for this bug then :-)

Thanks for the feedback!

**Comment by frenzibyte** [comment]
> > should just be a `textUpdateSchedule.Add(` -> `ScheduleAfterChildren(`
> 
> Just tried this locally, it doesn't cause any test failures and for the most part textboxes still behave the same, however the caret positioning in the case of the overriding textbox gets broken by this change (the caret displays behind the 1st character after the text override, instead of at its original position, it fixes itself if you move it though).

Oh yeah forgot to mention, the cursor and layout invalidation should now be inside the scheduler as that's where the caret position is modified, that should fix this issue:
image

**Comment by smoogipoo** [comment]
I'm weirded out that we update `Current` only in `UpdateAfterChildren` and enforce consistency only when the `TextBox` is not loaded. I would expect `Current` and `Text` to exactly mirror each other regardless of load state.

Is there a reason why we don't update it straight away? Seems like the following should work:

```diff
diff --git a/osu.Framework/Graphics/UserInterface/TextBox.cs b/osu.Framework/Graphics/UserInterface/TextBox.cs
index 9ca0e437d..2bb03e179 100644
--- a/osu.Framework/Graphics/UserInterface/TextBox.cs
+++ b/osu.Framework/Graphics/UserInterface/TextBox.cs
@@ -118,7 +118,7 @@ protected TextBox()
                 },
             };

-            Current.ValueChanged += e => { Text = e.NewValue; };
+            Current.ValueChanged += e => Text = e.NewValue;
             caret.Hide();
         }

@@ -319,9 +319,6 @@ private void updateCursorAndLayout()
             if (HasFocus)
                 caret.DisplayAt(new Vector2(cursorPos, 0), selectionWidth);

-            if (textAtLastLayout != text)
-                Current.Value = text;
-
             if (textAtLastLayout.Length == 0 || text.Length == 0)
             {
                 if (text.Length == 0)
@@ -340,8 +337,8 @@ protected override void UpdateAfterChildren()
             //have to run this after children flow
             if (!cursorAndLayout.IsValid)
             {
-                cursorAndLayout.Validate();
                 updateCursorAndLayout();
+                cursorAndLayout.Validate();
             }
         }

@@ -632,11 +629,13 @@ public virtual string Text
                 else
                     Placeholder.Hide();

-                if (!IsLoaded)
-                    Current.Value = text = value;
+                Current.Value = text = value;

                 textUpdateScheduler.Add(delegate
                 {
+                    if (text != value)
+                        return;
+
                     int startBefore = selectionStart;
                     selectionStart = selectionEnd = 0;
                     TextFlow?.Clear();
```

**Comment by bdach** [comment]
The above change passes the tests, and I *think* it makes some sense... The way I read it, adding the check ensures that the only scheduled delegate that does anything is the one with the exact same value of `text` as actually expected. Not 100% sure what it has to do with the `textAtLastLayout` check. Blame doesn't help as this goes way back and was in the game originally, so history is cut off partially...

**Comment by bdach** [comment]
In fact, if the test is modified as follows:

```diff
diff --git a/osu.Framework.Tests/Visual/UserInterface/TestSceneTextBox.cs b/osu.Framework.Tests/Visual/UserInterface/TestSceneTextBox.cs
index ab050bc2af..04b66726be 100644
--- a/osu.Framework.Tests/Visual/UserInterface/TestSceneTextBox.cs
+++ b/osu.Framework.Tests/Visual/UserInterface/TestSceneTextBox.cs
@@ -216,7 +216,9 @@ public void TestInputOverride()
             });
 
             AddStep(@"set some text", () => overrideInputBox.Text = "smth");
-            AddAssert(@"verify display state", () => overrideInputBox.FlowingText.Length == "Input overridden!".Length);
+            AddAssert(@"verify display state", () => overrideInputBox.FlowingText == "Input overridden!");
+            AddStep("append more text", () => overrideInputBox.InsertString("something else"));
+            AddAssert(@"verify display state", () => overrideInputBox.FlowingText == "Input overridden!");
         }
```

removing the `textAtLastLayout` makes it fail.

**Pull request closed without merging**